### PR TITLE
[WIN32K] Add StretchBlt function ability to flip images

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Fix StretchBlt function to allow flipped images

_Modify dib\dibxxbpp.c programs to understand flipped images._

JIRA issue: [CORE-16642](https://jira.reactos.org/browse/CORE-16642)
and
JIRA issue: [CORE-14408](https://jira.reactos.org/browse/CORE-14408)

Improvements to JIRA issue: [CORE-16634](https://jira.reactos.org/browse/CORE-16634)
 1) Mirroring Horizontally works.
 2) Mirroring Vertically works.
 3) Rotation 180° works.

StretchBlt_Fixes

See Videos at [CORE-16642](https://jira.reactos.org/browse/CORE-16642)

This is an update of [PR#2942](https://github.com/reactos/reactos/pull/2942)

Before:
ReactOS_StretchBlt_Test_Fail

After:
ReactOS_StretchBlt_Test_Fixed

### CODE REVIEW

File: win32ss/gdi/dib/dib24bpp.c [hunk 1]
```diff
@@ -53,18 +57,51 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+         BltInfo->DestSurface->sizlBitmap.cx, BltInfo->DestSurface->sizlBitmap.cy,
+         BltInfo->DestRect.left, BltInfo->DestRect.top, BltInfo->DestRect.right, BltInfo->DestRect.bottom);
+
+  /* Get back left to right flip here */
+  bLeftToRight = (BltInfo->DestRect.left > BltInfo->DestRect.right);
+
+  /* Check for top to bottom flip needed. */
+  bTopToBottom = BltInfo->DestRect.top > BltInfo->DestRect.bottom;
+
+  DPRINT("BltInfo->SourcePoint.x is '%d' & BltInfo->SourcePoint.y is '%d'.\n",
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:73]:
```suggestion
  DPRINT("BltInfo->SourcePoint.x is '%d' and BltInfo->SourcePoint.y is '%d'.\n",
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 2]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            }
+
+            sx = BltInfo->SourcePoint.x;
+            /* This sets sy to the top line */
+            sy = BltInfo->SourcePoint.y;
+
+            /* This sets sx to the rightmost pixel */
+            sx = BltInfo->SourcePoint.x + (BltInfo->DestRect.right - BltInfo->DestRect.left - 1);
+
+            for (j=BltInfo->DestRect.top; j<BltInfo->DestRect.bottom; j++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:332]:
```suggestion
            for (j = BltInfo->DestRect.top; j < BltInfo->DestRect.bottom; j++)
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 3]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            {
+
+              /* This sets sx to the rightmost pixel */
+              sx = BltInfo->SourcePoint.x + (BltInfo->DestRect.right - BltInfo->DestRect.left - 1);
+
+              Index = 0;
+
+              // Read right to left and store
+              for (i=BltInfo->DestRect.left; i<BltInfo->DestRect.right; i++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:341]:
```suggestion
              for (i = BltInfo->DestRect.left; i < BltInfo->DestRect.right; i++)
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 4]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+              {
+                store[Index] = DIB_24BPP_GetPixel(BltInfo->SourceSurface, sx, sy);
+                Index++;
+                sx--;
+              }
+
+              Index = 0;
+
+              // Write left to right to pixel
+              for (i=BltInfo->DestRect.left; i<BltInfo->DestRect.right; i++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:351]:
```suggestion
              for (i = BltInfo->DestRect.left; i < BltInfo->DestRect.right; i++)
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 5]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+              sx = BltInfo->SourcePoint.x;
+
+              /* This sets sy to the top line */
+              sy = BltInfo->SourcePoint.y;
+
+              /* This sets sy to the bottom line */
+              sy = BltInfo->SourcePoint.y + (BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1);
+            }
+
+            for (i=BltInfo->DestRect.left; i<BltInfo->DestRect.right; i++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:400]:
```suggestion
            for (i = BltInfo->DestRect.left; i < BltInfo->DestRect.right; i++)
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 6]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            {
+
+              /* This sets sy to the bottom line */
+              sy = BltInfo->SourcePoint.y + (BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1);
+              Index = 0;
+
+              /* Read bottom to top and store */
+              for (j=BltInfo->DestRect.top; j<BltInfo->DestRect.bottom; j++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:408]:
```suggestion
              for (j = BltInfo->DestRect.top; j < BltInfo->DestRect.bottom; j++)
```

File: win32ss/gdi/dib/dib24bpp.c [hunk 7]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+                {
+                  store[Index] = DIB_24BPP_GetPixel(BltInfo->SourceSurface, sx, sy);
+                }
+                Index++;
+                sy--;
+              }
+
+              Index = 0;
+
+              for (j=BltInfo->DestRect.top; j<BltInfo->DestRect.bottom; j++)
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:424]:
```suggestion
              for (j = BltInfo->DestRect.top; j < BltInfo->DestRect.bottom; j++)
```

File: win32ss/gdi/dib/dib32bpp.c [hunk 8]
```diff
@@ -51,46 +55,135 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+         BltInfo->DestSurface->sizlBitmap.cx, BltInfo->DestSurface->sizlBitmap.cy,
+         BltInfo->DestRect.left, BltInfo->DestRect.top, BltInfo->DestRect.right, BltInfo->DestRect.bottom);
+
+  DPRINT("BltInfo->DestSurface->lDelta is '%d' and BltInfo->SourceSurface->lDelta is '%d'.\n",
+           BltInfo->DestSurface->lDelta,  BltInfo->SourceSurface->lDelta);
+
+  DPRINT("iBitmapFormat is %d and width,height is (%d,%d).\n", BltInfo->SourceSurface->iBitmapFormat,
+         BltInfo->DestRect.right - BltInfo->DestRect.left, BltInfo->DestRect.bottom - BltInfo->DestRect.top);
+
+  DPRINT("BltInfo->SourcePoint.x is '%d' & BltInfo->SourcePoint.y is '%d'.\n",
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:79]:
```suggestion
  DPRINT("BltInfo->SourcePoint.x is '%d' and BltInfo->SourcePoint.y is '%d'.\n",
```

File: win32ss/gdi/dib/dib32bpp.c [hunk 9]
```diff
@@ -197,82 +364,284 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
-          DestBits += BltInfo->DestSurface->lDelta;
         }
       }
       else
       {
-        SourceBits = (PBYTE)BltInfo->SourceSurface->pvScan0 + ((BltInfo->SourcePoint.y + BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1) * BltInfo->SourceSurface->lDelta) + 4 * BltInfo->SourcePoint.x;
-        DestBits = (PBYTE)BltInfo->DestSurface->pvScan0 + ((BltInfo->DestRect.bottom - 1) * BltInfo->DestSurface->lDelta) + 4 * BltInfo->DestRect.left;
-        for (j = BltInfo->DestRect.bottom - 1; BltInfo->DestRect.top <= j; j--)
+        /* Buffering for source and destination flip overlaps. Fixes KHMZ MirrorTest CORE-16642 */
+        BOOL OneDone = FALSE;
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:499]: `OneDone` is too vague name. `NeedBuffer` or what?

File: win32ss/gdi/dib/dib32bpp.c [hunk 10]
```diff
@@ -197,82 +364,284 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+
+              Index = 0;
+
+              /* Copy stored dat to pixels from right to left */
+              for (i = BltInfo->DestRect.right - 1; BltInfo->DestRect.left <= i; i--)
+              {
+                *Dest32-- = store[Index];
+                Index++;
+              }
+            SourceBits -= BltInfo->SourceSurface->lDelta;
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:549]: Indent?

File: win32ss/gdi/dib/dib32bpp.c [hunk 11]
```diff
@@ -197,82 +364,284 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            SourceBitsT = (PBYTE)BltInfo->SourceSurface->pvScan0
+              + (BltInfo->SourcePoint.y * BltInfo->SourceSurface->lDelta) + 4 * BltInfo->SourcePoint.x;
           }
-          SourceBits -= BltInfo->SourceSurface->lDelta;
-          DestBits -= BltInfo->DestSurface->lDelta;
+
+          for (j = 0; j < (BltInfo->DestRect.bottom - BltInfo->DestRect.top) / 2 ; j++)
+          {
+            /* Store bottom row of Source pixels */
+            RtlMoveMemory(&store[0], SourceBitsB, 4 * (BltInfo->DestRect.right - BltInfo->DestRect.left));
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:616]:
```suggestion
            RtlMoveMemory(store, SourceBitsB, 4 * (BltInfo->DestRect.right - BltInfo->DestRect.left));
```

File: win32ss/gdi/dib/dib32bpp.c [hunk 12]
```diff
@@ -197,82 +364,284 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+
+            /* Copy top Source row to bottom Destination row overwriting it */
+            RtlMoveMemory(DestBitsB, SourceBitsT, 4 * (BltInfo->DestRect.right - BltInfo->DestRect.left));
+
+            /* Copy stored bottom row of Source pixels to Destination top row of pixels */
+            RtlMoveMemory(DestBitsT, &store[0], 4 * (BltInfo->DestRect.right - BltInfo->DestRect.left));
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:622]:
```suggestion
            RtlMoveMemory(DestBitsT, store, 4 * (BltInfo->DestRect.right - BltInfo->DestRect.left));
```

File: win32ss/gdi/dib/stretchblt.c [hunk 13]
```diff
@@ -118,15 +148,31 @@ BOOLEAN DIB_XXBPP_StretchBlt(SURFOBJ *DestSurf, SURFOBJ *SourceSurf, SURFOBJ *Ma
...
     for (DesX = DestRect->left; DesX < DestRect->right; DesX++)
     {
       CanDraw = TRUE;
 
       if (fnMask_GetPixel)
       {
-        sx = SourceRect->left+(DesX - DestRect->left) * SrcWidth / DstWidth;
+        if (bLeftToRight)
+        {
+          sx = SourceRect->right-(DesX - DestRect->left) * SrcWidth / DstWidth;  // flips about the y-axis
```
Comment (katahiromz) [inline win32ss/gdi/dib/stretchblt.c:170]:
```suggestion
          sx = SourceRect->right - (DesX - DestRect->left) * SrcWidth / DstWidth;  // flips about the y-axis
```

File: win32ss/gdi/dib/stretchblt.c [hunk 14]
```diff
@@ -137,7 +183,14 @@ BOOLEAN DIB_XXBPP_StretchBlt(SURFOBJ *DestSurf, SURFOBJ *SourceSurf, SURFOBJ *Ma
...
       if (UsesSource && CanDraw)
       {
-        sx = SourceRect->left+(DesX - DestRect->left) * SrcWidth / DstWidth;
+        if (bLeftToRight)
+        {
+          sx = SourceRect->right-(DesX - DestRect->left) * SrcWidth / DstWidth;  // flips about the y-axis
+        }
+        else
+        {
+          sx = SourceRect->left+(DesX - DestRect->left) * SrcWidth / DstWidth;
```
Comment (katahiromz) [inline win32ss/gdi/dib/stretchblt.c:192]:
```suggestion
          sx = SourceRect->left + (DesX - DestRect->left) * SrcWidth / DstWidth;
```

File: win32ss/gdi/eng/stretchblt.c [hunk 15]
```diff
@@ -415,6 +629,88 @@ IntEngStretchBlt(SURFOBJ *psoDest,
...
     }
 
+    DPRINT("source and dest size are NOT equal.\n");
+
+    DPRINT("SourceRect: (%d,%d)-(%d,%d) and DestRect: (%d,%d)-(%d,%d)\n",
+           SourceRect->left, SourceRect->top, SourceRect->right, SourceRect->bottom,
+           DestRect->left, DestRect->top, DestRect->right, DestRect->bottom);
+
+    /* if cxSrc < 0 then we change the signs for both cxSrc and cxDest and */
+    /* we reverse their coordinates, because these outcomes are the same.  */
```
Comment (katahiromz) [inline win32ss/gdi/eng/stretchblt.c:639]: Unite comments.

File: win32ss/gdi/eng/stretchblt.c [hunk 16]
```diff
@@ -415,6 +629,88 @@ IntEngStretchBlt(SURFOBJ *psoDest,
...
+        SourceRect->left = SourceRect->right;
+        SourceRect->right = lTmp;
+        lTmp = DestRect->left;
+        DestRect->left = DestRect->right;
+        DestRect->right = lTmp;
+        cxSrc = -cxSrc;
+        cxDest = -cxDest;
+    }
+    /* if cySrc < 0 then we change the signs for both cySrc and cyDest and
+     * we reverse their coordinates, because these outcomes are the same. */
```
Comment (katahiromz) [inline win32ss/gdi/eng/stretchblt.c:652]: Unite comments.

File: win32ss/gdi/eng/stretchblt.c [hunk 17]
```diff
@@ -429,24 +725,36 @@ IntEngStretchBlt(SURFOBJ *psoDest,
...
     if (InputRect.right < InputRect.left ||
             InputRect.bottom < InputRect.top)
     {
+        DPRINT("Returning TRUE.\n");
         /* Everything clipped away, nothing to do */
         return TRUE;
     }
 
+    DPRINT("InputRect: (%d,%d)-(%d,%d) and InputClippedRect: (%d,%d)-(%d,%d)\n",
+                InputRect.left, InputRect.top, InputRect.right, InputRect.bottom,
```
Comment (katahiromz) [inline win32ss/gdi/eng/stretchblt.c:742]: Indent?

File: win32ss/gdi/dib/dib16bpp.c [hunk 18]
```diff
@@ -330,21 +432,178 @@ DIB_16BPP_BitBltSrcCopy(PBLTINFO BltInfo)
           DestLine -= BltInfo->DestSurface->lDelta;
         }
       }
+    }
+      else
+      {
+        /* Buffering for source and destination flip overlaps. Fixes KHMZ MirrorTest CORE-16642 */
+        BOOL OneDone = FALSE;
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib16bpp.c:439]: OneDone is too vague name. Flipper or something better name...

File: win32ss/gdi/dib/dib16bpp.c [hunk 19]
```diff
@@ -330,21 +432,178 @@ DIB_16BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            * BltInfo->SourceSurface->lDelta) + 2 * BltInfo->SourcePoint.x;
+
+          /* Sets DestBits to the bottom line */
+          DestBits = (PBYTE)BltInfo->DestSurface->pvScan0
+           + (BltInfo->DestRect.bottom - 1) * BltInfo->DestSurface->lDelta
+           + 2 * BltInfo->DestRect.left + 2;
+
+          for (j = BltInfo->DestRect.bottom - 1; BltInfo->DestRect.top <= j; j--)
+          {
+
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib16bpp.c:467]: Delete this empty line.

File: win32ss/gdi/dib/dib24bpp.c [hunk 20]
```diff
@@ -191,28 +302,172 @@ DIB_24BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
+            /* Allocate enough pixels for a column in DWORD's */
+            DWORD *store = ExAllocatePoolWithTag(NonPagedPool,
+              (BltInfo->DestRect.bottom - BltInfo->DestRect.top + 1) * 4, TAG_DIB);
+            if (store == NULL)
+            {
+              DPRINT1("Storage Allocation Failed.\n");
+              return FALSE;
+            }
+
+            /* The OneDone flag indicates that we are flipping for bTopToBottom and bLeftToRight   */
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib24bpp.c:376]: Unite comments.

File: win32ss/gdi/dib/dib32bpp.c [hunk 21]
```diff
@@ -51,46 +55,135 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
...
   case BMF_1BPP:
+    DPRINT("1BPP Case Selected with DestRect Width of '%d'.\n",
+           BltInfo->DestRect.right - BltInfo->DestRect.left);
+
+    if (bLeftToRight || bTopToBottom)
+    DPRINT("bLeftToRight is '%d' and bTopToBottom is '%d'.\n", bLeftToRight, bTopToBottom);
 
     sx = BltInfo->SourcePoint.x;
+
+   /* This sets sy to the top line */
```
Comment (katahiromz) [inline win32ss/gdi/dib/dib32bpp.c:122]: Indent?

File: win32ss/gdi/dib/stretchblt.c (hunk 13 above)
Comment (Doug-Lyons) [reply win32ss/gdi/dib/stretchblt.c:170]: Fixed.

**Review by Doug-Lyons — COMMENTED** [review]
All Comments should now be addressed. Thanks.

**Comment by Doug-Lyons** [comment]
Thanks lots, @katahiromz. Its great to see this committed. Doug Lyons

**Pull request merged**